src/protocol/get_pfm_id.rs

```diff
@@ -0,0 +1,214 @@
+    /// order to parse the response. As of now, we will be diverging from that.
+    /// Instead, we encode and decode this with an extra byte that indicates
+    /// whether the response contains an `IdentifierType::Version` or
+    /// `IdentifierType::Platform` and the encoding of this follows that of the
+    /// request.
+    ///
+    /// The encoding is as follows:
+    ///
+    /// 1 - PFMValid (0 or 1)
+    /// 2 - IdentifierType (as encoded in the request)
+    /// 3:N - Identifier
+    ///       - If `Version` this will be exactly 4 bytes
+    ///       - If `Platform` this will be an ASCII string (not null terminated).
+    type GetPFMId;
+    const TYPE: CommandType = GetPFMId;
+
+    struct Request {
+        /// The port of the device whose PFM Id is being looked up.
+        pub port_id: u8,
+        /// The PFM Region (Active or Pending).
+        pub pfm_region: PFMRegion,
+        /// The ID Type (Version or Platform).
+        pub identifier: IdentifierType,
+    }
+
+    fn Request::from_wire(r, arena) {
+        let port_id = r.read_le()?;
+        let pfm_region = PFMRegion::from_wire(r, arena)?;
+        let identifier = if r.remaining_data() > 0 {
+            IdentifierType::from_wire(r, arena)?
+        } else {
+            // If missing, default identifier is `Version`
+            IdentifierType::Version
+        };
+        Ok(Self {
+            port_id,
+            pfm_region,
+            identifier,
+        })
+    }
+
+    fn Request::to_wire(&self, w) {
+        w.write_le(self.port_id)?;
+        self.pfm_region.to_wire(&mut w)?;
+        // TODO(asnaider): If `IdentifierType::Version` do we skip to save bytes?
+        self.identifier.to_wire(&mut w)?;
+        Ok(())
+    }
+
+    struct Response<'wire> {
+        /// PFM validity.
+        pub pfm_valid: PFMValid,
+        /// PFM Identifier (either Version or Platform).
+        #[cfg_attr(feature = "serde", serde(borrow))]
+        pub pfm_id: Identifier<'wire>,
+    }
+
+    fn Response::from_wire(r, arena) {
+        let pfm_valid = PFMValid::from_wire(r, arena)?;
+        let identifier = IdentifierType::from_wire(r, arena)?;
+        let pfm_id = match identifier {
+            IdentifierType::Version => {
+                let mut version: [u8; 4] = [0; 4];
+                r.read_bytes(&mut version)?;
+                Identifier::Version {id: version}
+            }
+            IdentifierType::Platform => {
+                if r.remaining_data() == 0 {
+                    Identifier::Platform {id: &[]}
+                } else {
+                    let bytes = r.read_direct(
+                    arena,
+                    Layout::from_size_align(
+                        r.remaining_data(),
+                        core::mem::align_of::<u8>())
+                    .expect(
+                        "Layout for u8 slice can't be created. This indicates a bug in the code."))?;
```

Comment:
Use `r.read_slice()` from `ReadZeroExt` instead. Make sure to add
```rust
use crate::io::ReadZeroExt as _;
```
to the imports.